### DISCUSSION TITLE
Cross-attention optimizations

### DISCUSSION THREAD

**Discussion by vladmandic** [vladmandic/sdnext#109] [Announcements]
As of recently, I've moved all command line flags regarding cross-optimization options to UI settings  
So things like `--xformers` are gone  

Default method is **scaled dot product** from `torch` 2.0  
And its probably best unless you're running on low-powered GPU (e.g. nVidia 1xxx), in which case `xformers` are still better  

Note that `xformers` are not released in binary form on Windows for Torch 2.0 and CUDA 11.8. So at the moment, only way to get `xformers` on Windows is to either manually compile them (requires compiler to be installed) or to find a 3rd party compiled ones (*cannot guarantee quality of or any 3rd party compiled code*)  

And AMD users report that **sub-quad attention** achieves best results on some systems, so give that a try

**Comment by mart-hill** [comment]
I noticed that Tiled VAE/Tiled Diffusion doesn't want to work, unless I turn on xformers (which I installed along with Torch 2.0). Am I missing something within the config of the UI maybe?

**Reply by vladmandic** [reply]
tiled vae doesn't work with sdp? which platform are you on?
i haven't used that feature in a while, i'll check (but i doubt its a config issue).

**Reply by mart-hill** [reply]
I'm on Win10 22H2 with RTX 3090.
I had specific "framed" error message with xformers at the end, when I tried to use Tiled VAE with scaled dot product. The error made me aware, that I probably should enable xformers in the configuration, and then restart the UI entirely, which worked.

Edit: I corrected the mistake and Tiled VAE works with SDP now for me. I'll check, if it works with Invoke AI cross-attention optimization method.

**Reply by razvan-nicolae** [reply]
@vladmandic are you aware of the issues similar to this thread ? https://www.reddit.com/r/StableDiffusion/comments/12grqts/cuda_out_of_memory_errors_after_upgrading_to/
A lot of people on torch 2.0 are seeing this, myself included, txt2img raw batch VRAM is severely degraded with this version, less than 50% of what old torch/automatic1111 was able to support with the same hardware ( 4090 in my case - out of VRAM at 3 images of around 1000x1000 pixels, which is pretty much lol)

**Reply by vladmandic** [reply]
i'm actually seeing quite good results with torch 2.0 + sdp and some tweaks:

```python
    os.environ.setdefault('TF_CPP_MIN_LOG_LEVEL', '2')
    os.environ.setdefault('ACCELERATE', 'True')
    os.environ.setdefault('FORCE_CUDA', '1')
    os.environ.setdefault('ATTN_PRECISION', 'fp16')
    os.environ.setdefault('PYTORCH_CUDA_ALLOC_CONF', 'garbage_collection_threshold:0.9,max_split_size_mb:512')
    os.environ.setdefault('CUDA_LAUNCH_BLOCKING', '0')
    os.environ.setdefault('CUDA_CACHE_DISABLE', '0')
    os.environ.setdefault('CUDA_AUTO_BOOST', '1')
    os.environ.setdefault('CUDA_MODULE_LOADING', 'LAZY')
    os.environ.setdefault('CUDA_DEVICE_DEFAULT_PERSISTING_L2_CACHE_PERCENTAGE_LIMIT', '0')
    os.environ.setdefault('GRADIO_ANALYTICS_ENABLED', 'False')
    os.environ.setdefault('SAFETENSORS_FAST_GPU', '1')
    os.environ.setdefault('NUMEXPR_MAX_THREADS', '16')
```

**Reply by razvan-nicolae** [reply]
Just tested with your tweaks added to launch.py

![image](https://user-images.githubusercontent.com/19169014/231905022-96dff7b3-66a6-4984-9c52-39a4987b3bf8.png)

**Reply by razvan-nicolae** [reply]
my question is, where do people see "performance gains" with SDP over Xformers - maybe quality gains? but my xformers was very clean i am comparing seed by seed and i don't like the SDP results more than the xformers one so far.
Where are the "50%" faster results? Can anyone confirm with 4090?

**Reply by vladmandic** [reply]
i haven't said sdp is 50% faster than xformers. but its much cleaner implementation-wise, far less installation problems (xformers is quite often not available for this-or-that-platform in binary form and needs to be recompiled separately for any minor change to torch, etc.)

so if sdp "just works", i'm happy.
xformers were amazing, but now i'd recommend them only for much older gpu's which do not have sufficient pipeline. for example, nvidia 1xxx series is better with xformers. the rest? why bother, just my $0.02.

also, xformers also break quite a few things related to training, so need to unwind them each time, etc. end-user may not care, but from dev perspective, i hate doing that all the time.

**Reply by derspanier1** [reply]
@vladmandic from your exchange i gather your mentioned tweaks can be inserted into the launch.py ? i assume in the header section right ? i tested this, and it seems to be right, and working, but its gets removed with every upgrade. is there another way of inserting the tweaks without being deleted by an update ? Thank you for this fork, i absolutely love it.

**Reply by vladmandic** [reply]
If you're using my fork, those ops are already built-in.

**Reply by derspanier1** [reply]
@vladmandic thank you for explaining, appreciated. yeah using your fork, tested it 2 days ago, ran so much better than main auto, migrated everything to this install. works and feels much much better. Thank you for your work !

**Reply by BSnowdonia** [reply]
You can try to update the cuDNN component to version 8.8.x. The version of this component can be checked by in `sd-extension-system-info`. The default built-in cuDNN version of pytorch 2.0 is 8500, and in order to maximize the performance of the 40xx card, you need to update it to version 88xx, which can be found at [cuDNN](https://developer.nvidia.com/rdp/cudnn-download). placing the cudnn*.dll in the `bin` folder of the zip package into the `%PYTHON_ROOT%\Lib\site-packages\torch\lib` directory, where `%PYTHON_ROOT%` is the path of your python installation.
These are some of the measurement results from which you can find some differences.
![screencap](https://user-images.githubusercontent.com/70185062/234486386-1abb9aed-5b8c-478b-a600-e53881331528.png)
This method also has some efficiency gains for other models of graphics cards, so give it a try.

**Comment by razvan-nicolae** [comment]
No, I didn’t say you suggested that, but it was posted in a few places, one big study here

<https://pytorch.org/blog/accelerated-diffusers-pt-20/>
[pytorch-logo.png]
PyTorch<https://pytorch.org/blog/accelerated-diffusers-pt-20/>
pytorch.org<https://pytorch.org/blog/accelerated-diffusers-pt-20/>



The pytotch devs are claiming these improvements, but honestly I don’t see them.

P.S: offtopic is there any way to customize font in your fork? Eyes are getting tired of the boldish looking default font :D



On 14 Apr 2023, at 02:53, Vladimir Mandic ***@***.***> wrote:

﻿

i haven't said sdp is 50% faster than xformers. but its much cleaner implementation-wise, far less installation problems (xformers is quite often not available for this-or-that-platform in binary form and needs to be recompiled separately for any minor change to torch, etc.)

so if sdp "just works", i'm happy.
xformers were amazing, but now i'd recommend them only for much older gpu's which do not have sufficient pipeline. for example, nvidia 1xxx series is better with xformers. the rest? why bother, just my $0.02.

also, xformers also break quite a few things related to training, so need to unwind them each time, etc. end-user may not care, but from dev perspective, i hate doing that all the time.

—
Reply to this email directly, view it on GitHub<https://github.com/vladmandic/automatic/discussions/109#discussioncomment-5610227>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AESH55WR5YNAOZKTUYSF5ILXBCGVFANCNFSM6AAAAAAW5LKODY>.
You are receiving this because you commented.Message ID: ***@***.***>

_________________________________________________________________________________________________________________________

Ce message et ses pieces jointes peuvent contenir des informations confidentielles ou privilegiees et ne doivent donc
pas etre diffuses, exploites ou copies sans autorisation. Si vous avez recu ce message par erreur, veuillez le signaler
a l'expediteur et le detruire ainsi que les pieces jointes. Les messages electroniques etant susceptibles d'alteration,
Orange decline toute responsabilite si ce message a ete altere, deforme ou falsifie. Merci.

This message and its attachments may contain confidential or privileged information that may be protected by law;
they should not be distributed, used or copied without authorisation.
If you have received this email in error, please notify the sender and delete this message and its attachments.
As emails may be altered, Orange is not liable for messages that have been modified, changed or falsified.
Thank you.

**Comment by vladmandic** [comment]
> P.S: offtopic is there any way to customize font in your fork? Eyes are getting tired of the boldish looking default font :D

not cleanly yet, its on my todo. but for now, you can edit `user.css`.

**Comment by TomerGadol** [comment]
When you say Xformers is better than SDP on old GPUs, like my 1660ti, is the difference noticeable? Are there any measurements?
If not, I might just do a few tests myself.

**Reply by vladmandic** [reply]
> When you say Xformers is better than SDP on old GPUs, like my 1660ti, is the difference noticeable? Are there any measurements? If not, I might just do a few tests myself.

nearly double. reason is that sdp performs most ops on gpu while xformers performs some on on cpu.
so if you have modern gpu, having it all done on gpu is ok.
and if you have a high end gpu and cpu with not extreme clocks (e.g. below 5ghz), sdp is actually better as cpu in those cases could not keep gpu saturated.
but if you have old gpu like 1660, xformers help by doing some work on cpu so gpu focuses on core work.

actual work done by sdp and xformers is pretty comparable, its just where its done - so it depends on the system which is better.

**Reply by TomerGadol** [reply]
Ok, thanks for the explanation. Do you know how `--medvarm` and `--no-half` affect performance with SDP and xformers? On A1111 I needed both arguments with xformers enabled to get anything bigger than 512x512. Is it possible to improve performance on my 1660ti by removing one of these and using SDP?

**Reply by vladmandic** [reply]
`--no-half` should not be needed for bigger resolutions, it should be needed for specific models only that are trained in fp32 and tend to produce zeros at fp16 precision. this was typically the case with sd2.0 original model (sd2.1 is better). symptoms are no runtime errors, just black image as output. also needed on some very specific (broken) gpus that handle fp16 badly. so unless you're getting black images, you don't have to use it and its going to be faster without it.
`--medvram` has no impact on actual performance once everything is running. it has impact what is loaded to vram vs ram. so startup is slower, but performance is the same.
`--lowvram` is much and does the same thing, but much more aggressively so it can cause paging vram<->ram during actual generation. so it will be sometimes fast and sometimes just stop until it pages.

**Reply by gunnarsal** [reply]
I have 1660 TI and with A1111 everything is fine. With vlad though it gives me only black images. I have tried to add "set COMMANDLINE_ARGS=--lowvram --precision full --no-half" to the webui.bat but it is either giving me an error on launc or does not do anything. 

Am I adding it to an incorrect spot?

**Reply by TomerGadol** [reply]
I've noticed a similar behavior: with A1111 I do not need `--no-half` yet with Vlad's I do need it to produce images. I'm not sure if it's something with the implementation of xformers or something else since initially I needed it also with A1111 and only after some updates could I remove it. 
I do think you can remove the precision flag tho, but you might be missing an "-' between '--precision' to 'full' -> `-precision-full`.
I'm using only `--medvram --no-half`

**Reply by gunnarsal** [reply]
Thanks for quick reply but where do you actually add this? in what spot inside the webui.bat or do you add it somewhere in the settings when Vlad is already running?

And do I need to add this beginning then in the notpad as well before them? COMMANDLINE_ARGS=

**Reply by mart-hill** [reply]
With @vladmandic's fork I only use --no-half-vae, everything else is set now by Settings, and because API is exposed by default, it's useful with Chainner for example. 🙂

**Reply by gunnarsal** [reply]
> With @vladmandic's fork I only use --no-half-vae, everything else is set now by Settings, and because API is exposed by default, it's useful with Chainner for example. 🙂

So you edit the webui.bat with notepad and add COMMANDLINE_ARGS=--no-half-vae there? But between what text? I cant get it to recognize or atleast it still gives me black images afterwards

**Reply by Shadowghost** [reply]
You should be able to fix your problems by enable `upcast sampling` (if you require `--no-half`) in the `Stable Diffusion` setting and and `Use full precision for VAE (--no-half-vae)` in the CUDA settings within the webui settings (upcast sampling is faster and less memory hungry than `--no-half`).
One of the main points of this fork is to move from CLI flags to settings in the UI.

**Reply by mart-hill** [reply]
I made webui-user.bat specifically with this content:

```
@echo off

set TMP=X:\AI\TEMP
set TEMP=X:\AI\TEMP
set PYTHON=
set GIT=
set VENV_DIR=
set COMMANDLINE_ARGS=--no-half-vae
set SAFETENSORS_FAST_GPU=1

call webui.bat
```

I start it from non-admin "perspective", hence the filename. 🙂

**Reply by gunnarsal** [reply]
Hmm, i have s***loads of other stuff in the webui.bat file as well for some reason, can I get rid of most/all of it?

`@echo off

if not defined PYTHON (set PYTHON=python)
if not defined VENV_DIR (set "VENV_DIR=%~dp0%venv")

set ERROR_REPORTING=FALSE

mkdir tmp 2>NUL

%PYTHON% -c "" >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :check_pip
echo Couldn't launch python
goto :show_stdout_stderr

:check_pip
%PYTHON% -mpip --help >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :start_venv
if "%PIP_INSTALLER_LOCATION%" == "" goto :show_stdout_stderr
%PYTHON% "%PIP_INSTALLER_LOCATION%" >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :start_venv
echo Couldn't install pip
goto :show_stdout_stderr

:start_venv
if ["%VENV_DIR%"] == ["-"] goto :skip_venv
if ["%SKIP_VENV%"] == ["1"] goto :skip_venv

dir "%VENV_DIR%\Scripts\Python.exe" >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :activate_venv

for /f "delims=" %%i in ('CALL %PYTHON% -c "import sys; print(sys.executable)"') do set PYTHON_FULLNAME="%%i"
echo Creating venv in directory %VENV_DIR% using python %PYTHON_FULLNAME%
%PYTHON_FULLNAME% -m venv "%VENV_DIR%" >tmp/stdout.txt 2>tmp/stderr.txt
if %ERRORLEVEL% == 0 goto :activate_venv
echo Unable to create venv in directory "%VENV_DIR%"
goto :show_stdout_stderr

:activate_venv
set PYTHON="%VENV_DIR%\Scripts\Python.exe"

echo venv %PYTHON%

:skip_venv
if [%ACCELERATE%] == ["True"] goto :accelerate
goto :launch

:accelerate
echo Checking for accelerate
set ACCELERATE="%VENV_DIR%\Scripts\accelerate.exe"
if EXIST %ACCELERATE% goto :accelerate_launch

:launch
%PYTHON% launch.py %*
pause
exit /b

:accelerate_launch
echo Accelerating
%ACCELERATE% launch --num_cpu_threads_per_process=6 launch.py
pause
exit /b

:show_stdout_stderr

echo.
echo exit code: %errorlevel%

for /f %%i in ("tmp\stdout.txt") do set size=%%~zi
if %size% equ 0 goto :show_stderr
echo.
echo stdout:
type tmp\stdout.txt

:show_stderr
for /f %%i in ("tmp\stderr.txt") do set size=%%~zi
if %size% equ 0 goto :show_stderr
echo.
echo stderr:
type tmp\stderr.txt

:endofscript

echo.
echo Launch unsuccessful. Exiting.
pause





`

**Reply by mart-hill** [reply]
By the way, this line:

`set SAFETENSORS_FAST_GPU=1`

means "by default models are loaded to CPU and then moved to GPU. SAFETENSORS has an untested option to load directly to GPU, thus bypassing one memory copy step"

**Reply by mart-hill** [reply]
For me, `webui.bat` is untouched, I only created this `webui-user.bat`, to redirect some "stuff" like temporary files for the cmd session and command line argument sent to the UI. Well, and this untested env variable. This "webui-user.bat" file name is ignored by `git pull`, so it's fine! 🙂

**Reply by vladmandic** [reply]
if you're using my fork, options like --no-half and similar will be part of ui settings shortly, no need to play with command line.

**Reply by TomerGadol** [reply]
I am using`Upcast cross attention layer to float32`, `Enable upcast sampling` and `Use full precision for VAE (--no-half-vae)` in CUDA settings but still get `NansException: A tensor with all NaNs was produced in VAE. This could be because there's not enough precision to
represent the picture. Try adding --no-half-vae commandline argument to fix this. Use --disable-nan-check commandline
argument to disable this check.`

**Reply by vladmandic** [reply]
if model if finetuned to produce extremely low values, it will result in NaN during math operations. `--no-half` reduces the problem by using 32bit precision instead of 16bit, but it doesn't solve the problem where some models are just bad.
and yes, torch 2.0 is more sensitive.

**Reply by TomerGadol** [reply]
This is happening with the base 1.5 model, enabled both upcast options in SD settings as well as both full precision options in CUDA settings. Default VAE set to 'none'

**Reply by vladmandic** [reply]
> This is happening with the base 1.5 model, enabled both upcast options in SD settings as well as both full precision options in CUDA settings. Default VAE set to 'none'

now, THAT should not be happening. i know nvidia 1xxx series had some internal issues enabling/disabled fp16/fp32 and there was a trick to it, but for the sake of me, cant remember the details. i think it had to do with enabling `cudnn.benchmark` and then voila, it work, although its totally unrelated.

**Reply by TomerGadol** [reply]
Should I open an issue for this? or just fall back to `--no-half`?

**Reply by mart-hill** [reply]
I also just encountered NaN with Torch 2.0 (never had these before with RTX 3090) Version: df424d6d Thu Apr 20 17:18:31 2023 -0400 of the UI. Since there's an issue with Tiled VAE addon, again (this time probably because it wants 32bits, but gets halfs; checked various options with newly created CUDA section in settings, no dime), I reopened the #128 and reported all the issued I've noticed.
For now, to avoid NaN in VAE, vae-ft-mse-840000-ema-pruned should be used, right? It's weird though, as I had preview of the image all the way to the end with "blessed.vae.pt", which is a variant of NAI VAE.

**Reply by nekoworkshop** [reply]
@mart-hill That is a known issue with NAI VAE or any VAE derived from the NAI VAE. You need `--no-half-vae`, not `--no-half`. The downside is a VRAM usage spike in the very last step.

The preview images are derived from the latent space using approximation, hence why they are still good. The NaN happens in the last step where the latent space is decoded using the VAE.

**Reply by mart-hill** [reply]
Exactly, though I had this VERY rarely, and had --no-half-vae all the time ON. But now this option landed in the Settings (it's not available as cmdline option anymore for this fork), and this probably needs a tiny bit of polishing. 🙂

**Comment by FurkanGozukara** [comment]
--xformers still works better than opt sdp attention
For example xformers can generate bigger resolution image and faster for older cards
Also in dreambooth training xformers was much faster
You may say how do you know I made extensive tests on this video



[**RTX 3090 vs RTX 3060 Ultimate Showdown for Stable Diffusion, ML, AI & Video Rendering Performance**](https://youtu.be/lgP1LNnaUaQ)
[![image](https://user-images.githubusercontent.com/19240467/231303430-63d801cf-3c5a-4c20-b445-bb682febfa4e.png)](https://youtu.be/lgP1LNnaUaQ)


**you really should support --xformers along with opt sdp attention**

**Reply by vladmandic** [reply]
and who says i don't? i just set sdp as default and i don't plan to remove xformers. and nothing you said is any different than what i said earlier?

**Reply by xedis** [reply]
> **you really should support --xformers along with opt sdp attention**

vlad made it part of the ui. quick ctrl+f  of the settings options after clicking show all pages would have went far here.

![image](https://user-images.githubusercontent.com/4144634/232346423-40834ea0-f4c8-4f05-91c5-367a5d4b15ef.png)

**Reply by FurkanGozukara** [reply]
> and who says i don't? i just set sdp as default and i don't plan to remove xformers. and nothing you said is any different than what i said earlier?

when you said you removed from command line arguments i thought it was removed 

@xedis  nice

**Reply by vladmandic** [reply]
> I've moved all command line flags regarding cross-optimization options to UI settings

How was that not clear?

**Comment by radry** [comment]
I tried both xformers and SDP in vlad fork on a RTX 4070 but got the exact same speed, so why is SDP better?
(Yes, I verified that the correct one is used and also restarted the server in between switching, so it will apply the change in settings)

**Reply by vladmandic** [reply]
Did I say they were? They perform the same (except on very old HW) with much less compatibility/install issues.

**Reply by CapsAdmin** [reply]
it also works on AMD hardware

**Comment by DaveScream** [comment]
On my 3060 12gb windows 10 I cant get the same speed as in a1111.
A1111 torch 1.13 + cu117 + xformers I get 7.4it/s (optimized previews)

Vlad torch 2.0 + cu118 sdp or xformers - 6.7it/s (live previews optimized or turned off, no matter)

So it is not possible to reach same speeds, sadly

**Comment by CapsAdmin** [comment]
> And AMD users report that sub-quad attention achieves best results on some systems, so give that a try

On a 6900XT I found that sub-quad attention performs the worst but has the lowest vram usage. So kind of useful to quickly switch to (thanks to it being moved to options) if you want to render something high resolution to avoid OOM.
